packages/fast-colors/src/component-state-color-palette.ts

```diff
@@ -0,0 +1,165 @@
+import { ColorRGBA64 } from "./color-rgba-64";
+import { ColorPalette } from "./color-palette";
+import { hslToRGB, rgbToHSL, rgbToLuminanceLinear } from "./color-converters";
+import { ColorScale, ColorScaleStop } from "./color-scale";
+import { ColorHSL } from "./color-hsl";
+import { ColorInterpolationSpace } from "./color-interpolation";
+import { parseColorHexRGB } from "./parse-color";
+
+export interface ComponentStateColorPaletteConfig {
+    baseColor?: ColorRGBA64;
+    steps?: number;
+}
+
+export class ComponentStateColorPalette {
+    public static readonly defaultPaletteConfig: ComponentStateColorPaletteConfig = {
+        baseColor: parseColorHexRGB("#808080"),
+        steps: 94,
+    };
+
+    public palette: ColorRGBA64[] = [];
+    private readonly config: ComponentStateColorPaletteConfig;
+
+    constructor(config: ComponentStateColorPaletteConfig) {
+        this.config = Object.assign(
+            {},
+            ComponentStateColorPalette.defaultPaletteConfig,
+            config
+        );
+        this.palette = [];
+        this.regenPalettes();
+    }
+
+    private regenPalettes(): void {
+        let s: number = this.config.steps;
+        if (isNaN(this.config.steps) || this.config.steps < 3) {
+            s = 3;
+        }
+
+        // This palette is tuned to go as dark as differences between the levels can be perceived according to tests
+        // on numerous monitors in different conditions. Stay linear from white until this first cutoff.
+        const darkLum: number = 0.14;
+
+        // In the dark compression, this is the last luminance value before full black.
+        const darkestLum: number = 0.06;
+
+        // The Color for the luminance value above, placed on the ramp at it's normal position, so darker colors after
+        // it can be compressed.
+        const darkLumColor: ColorRGBA64 = new ColorRGBA64(darkLum, darkLum, darkLum, 1);
+
+        // The number of steps in the ramp that has been tuned for default use. This coincides with the size of the
+        // default ramp, but the palette could be generated with fewer steps to increase final contrast. This number
+        // should however stay the same.
+        const stepsForLuminanceRamp: number = 94;
+
+        // Create the reference, dark-compressed, grey palette, like:
+        // F------------------------------------------------------------------------------------D-------0
+        const r: ColorPalette = new ColorPalette({
+            ...ColorPalette.greyscalePaletteConfig,
+            baseColor: darkLumColor,
+            baseScalePosition: ((1 - darkLum) * 100) / stepsForLuminanceRamp,
+            steps: s,
+        });
+        const referencePalette: ColorRGBA64[] = r.palette;
+
+        // Find the requested base color on the adjusted luminance reference ramp.
+        // There is no _right_ way to desaturate a color, and both methods we've tested have value, so average them out.
```

Comment:
should this denaturation step be configurable? For instance, an option on the config?

Reply:
You mean in terms of which algorithm or method to use? I don’t think so. I think it can be encapsulated in the larger generation algorithm, but it’s possible we’ll tweak this so we end up with ramps that are more consistent in terms of perceived brightness between steps regardless of color. We’ve been discussing options for a v2, but it’s to make the design more robust rather than to achieve a particular color that has preference over another.